cmake/Modules/set_cpu_arch_flags.cmake

```diff
@@ -0,0 +1,84 @@
+#=============================================================================
+# Copyright 2022 NVIDIA Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+
+#------------------------------------------------------------------------------#
+# Architecture
+#------------------------------------------------------------------------------#
+if(BUILD_MARCH AND BUILD_MCPU)
+  message(FATAL_ERROR "BUILD_MARCH and BUILD_MCPU are incompatible")
+endif()
+
+function(set_cpu_arch_flags out_var)
+  # Try -march first. On platforms that don't support it, GCC will issue a hard
+  # error, so we'll know not to use it. Default is "native", but explicitly
+  # setting BUILD_MARCH="" disables use of the flag
+  if(BUILD_MARCH)
+    set(INTERNAL_BUILD_MARCH ${BUILD_MARCH})
+  elseif(NOT DEFINED BUILD_MARCH)
+    set(INTERNAL_BUILD_MARCH "native")
+  endif()
+
+  set(flags "")
+
+  include(CheckCXXCompilerFlag)
+  if(INTERNAL_BUILD_MARCH)
+    check_cxx_compiler_flag("-march=${INTERNAL_BUILD_MARCH}" COMPILER_SUPPORTS_MARCH)
+    if(COMPILER_SUPPORTS_MARCH)
+      list(APPEND flags "-march=${INTERNAL_BUILD_MARCH}")
+    elseif(BUILD_MARCH)
+      message(FATAL_ERROR "The flag -march=${INTERNAL_BUILD_MARCH} is not supported by the compiler")
```

Comment:
IIUC correctly, this only fatals if -march is not valid and BUILD_MARCH was explicitly specified by the user?  If so, the error message to the user might be clearer by saying "BUILD_MARCH=xyz requested, but -march= is not supported. Unsetting -DBUILD_MARCH=xyz will allow compiler to continue without the -march flag." so it's clear the flag is not required and what user options are for fixing it.

Reply:
I copied this logic [from Legion](https://gitlab.com/StanfordLegion/legion/-/blob/master/CMakeLists.txt#L122-182) with the assumption they have it right. Since this logic is run after we find (or add) Legion, there's two scenarios:

1. Legion wasn't found, so CPM added it. In this case, if `BUILD_MARCH` is bad, Legion will cause the `FATAL_ERROR`
2. If Legion was found, we use the logic here. I'm assuming that if Legion was found, it's because a dev passed `-DLegion_ROOT=`, and so already knows the right way to pass `-DBUILD_MARCH`?

Reply:
Gotcha. That makes sense. LGTM. I can't mark resolved.